### DISCUSSION TITLE
Benchmarking Appwrite on ARM

### DISCUSSION THREAD

**Discussion by kodumbeats** [appwrite/appwrite#1041] [Show and tell]
With the upcoming `0.8.0` release, Appwrite will support ARM CPUs, and we're curious to see how it performs on the wide variety of ARM devices out there. As a self-hosting advocate, I'm especially excited to run **Appwrite on the Raspberry Pi**!

If you have a Raspberry Pi laying around (or five, like I do), try setting up Appwrite on the `feat-465-add-arm-support` branch of `appwrite/appwrite`. I've [written a tutorial](https://gist.github.com/kodumbeats/befdc97ee6d04dcca3af80f2a9196f1f) on how I deployed Appwrite to a Raspberry Pi 4, and I'd love to see what kinds of benchmarks you're able to generate.

We have enjoyed benchmarking with the open-source tool [k6](https://github.com/k6io/k6), and I've included a brief benchmarking script in my tutorial to test against the database health endpoint. If you've generated benchmarks, please share the details (CPU, memory, disk, endpoints, etc)!

**Comment by mlucas-NU** [comment]
### k6 Stress Test
**Note:** I had issues setting up the certificate for this instance (see below), so I'm running `k6 run script.js --insecure-skip-tls-verify`
<img width="716" alt="Screen Shot 2021-03-27 at 12 33 22 PM" src="https://user-images.githubusercontent.com/15447878/112732684-58bfc900-8f33-11eb-83f8-9063c15cbd17.png">


### ARM Server Details
**Device:** [Pine Rockpro64](https://www.pine64.org/rockpro64/)
**OS:** Armbian 20.11.4 (5.9.14-rockchip64 kernel)
**CPU:** Rockchip RK3399 SOC, which contains the following:
 * Dual-core Cortex-A72 up to 2.0GHz CPU
 * Quad-core Cortex-A53 up to 1.5GHz CPU

**Memory:** 4GB LPDDR4
**Disk:** 37GB free (57GB partition on a 64GB eMMC drive)

### Notes:
* I did not disable clamav.
* Everything appears to be running normally. Previously, I was building ARM images for appwrite, smtp, mariadb, clamav, and telegraf. I checked the logs for each of these and they all appear to be working as expected (minus smtp, which was disabled in this branch's docker-compose.yml).
* I created a project, user, and team manually without issue and assigned the user to the team. Basic functionality seems good, but this was by no means an exhaustive test.
* Browsers don't trust the traefik cert. It's been a month or more since I configured the cert on my 0.7.0 instance (which was upgraded to 0.7.1 when that was released), so I forget how I got appwrite to sign a valid cert or if I just accepted the warning.
* I copied my .env values over from my 0.7.1 instance on the same server (not running any appwrite instances at the time). The previous version was generating valid certificates, but my `feat-465-add-arm-support` instance could not. Here's the output of appwrite-worker-certificates:
```
appwrite-worker-certificates    | Appwrite certificates worker v1 has started
appwrite-worker-certificates    | [notice] Starting worker 3adcee34600a:7:v1-certificates
appwrite-worker-certificates    | [notice] Starting work on (Job{v1-certificates} | ID: 88557229123a05c1a357a37d4f35444d | CertificatesV1 | [{"document":[],"domain":"appwrite.mymaskeddomain.com","validateTarget":false,"validateCNAME":false}])
appwrite-worker-certificates    | [critical] (Job{v1-certificates} | ID: 88557229123a05c1a357a37d4f35444d | CertificatesV1 | [{"document":[],"domain":"appwrite.mymaskeddomain.com,"validateTarget":false,"validateCNAME":false}]) has failed Exception: Failed to issue a certificate with message: Saving debug log to /var/log/letsencrypt/letsencrypt.log
appwrite-worker-certificates    | Plugins selected: Authenticator webroot, Installer None
appwrite-worker-certificates    | Cert not yet due for renewal
appwrite-worker-certificates    | Keeping the existing certificate
appwrite-worker-certificates    |  in /usr/src/code/app/workers/certificates.php:128
appwrite-worker-certificates    | Stack trace:
appwrite-worker-certificates    | #0 /usr/src/code/vendor/resque/php-resque/lib/Resque/Job.php(201): CertificatesV1->perform()
appwrite-worker-certificates    | #1 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(303): Resque_Job->perform()
appwrite-worker-certificates    | #2 /usr/src/code/vendor/resque/php-resque/lib/Resque/Worker.php(242): Resque_Worker->perform(Object(Resque_Job))
appwrite-worker-certificates    | #3 /usr/src/code/vendor/resque/php-resque/bin/resque(185): Resque_Worker->work(5, false)
appwrite-worker-certificates    | #4 {main}
```

**Reply by kodumbeats** [reply]
If LetsEncrypt's ACME servers can't reach your SBC on ports 80/443, then you can't get an SSL cert signed. However, those errors remind me of this relevant issue on my to-do list: https://github.com/appwrite/appwrite/issues/1025

**Comment by vsantele** [comment]
![image](https://user-images.githubusercontent.com/26800140/112734936-13a69180-8f49-11eb-8c5c-95f6fa19c1ee.png)
Device: [nanopi M4](https://www.friendlyarm.com/index.php?route=product/product&product_id=234) 
OS: Ubuntu 20.04 (kernel: 4.4.167)
SoC: RK3399
CPU
- 2 Cortex-A72
- 4 Cortex-A53
Memory: 2GB DDR3

**Comment by kodumbeats** [comment]
### Writing documents to DB (~1600 written)

![write-to-rpi-4-appwrite](https://user-images.githubusercontent.com/9708641/112762970-75f6a500-8fd0-11eb-8460-5be18f8e0410.png)

Device: Raspberry Pi 4 Model B, 2GB RAM
Processor: Quad core Cortex-A72 @1.5GHz
Storage: [Samsung EVO Select 32GB SD Card](https://www.samsung.com/us/computing/memory-storage/memory-cards/microsdhc-evo-select-memory-card-w--adapter-32gb--2017-model--mb-me32ga-am/)

<details>
<summary>dbwrite.js</summary>
https://gist.github.com/kodumbeats/60e94e7d3a031f6084e7999288998ad6
</details>

**Comment by kodumbeats** [comment]
### Fetching a list of 1600 documents

![fetching1600docsonpi4](https://user-images.githubusercontent.com/9708641/112763649-51e89300-8fd3-11eb-912d-de7e3fe4bf20.png)

Device: Raspberry Pi 4 Model B, 2GB RAM
Processor: Quad core Cortex-A72 @1.5GHz
Storage: Samsung EVO Select 32GB SD Card

<details><summary>dbread.js</summary>
https://gist.github.com/kodumbeats/33a07e1c2e87e4c86e945c903d440fb8
</details>
